src/cpp/session/modules/rmarkdown/NotebookDocQueue.cpp

```diff
@@ -0,0 +1,190 @@
+/*
+ * NotebookDocQueue.cpp
+ *
+ * Copyright (C) 2009-16 by RStudio, Inc.
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+#include "NotebookDocQueue.hpp"
+
+#include <boost/foreach.hpp>
+
+#include <session/SessionModuleContext.hpp>
+
+using namespace rstudio::core;
+
+#define kDocQueueId             "doc_id"
+#define kDocQueueJobDesc        "job_desc"
+#define kDocQueuePixelWidth     "pixel_width"
+#define kDocQueueCharWidth      "pixel_width"
+#define kDocQueueUnits          "units"
+#define kDocQueueMaxUnits       "max_units"
+#define kDocQueueCommitMode     "commit_mode"
+#define kDocQueueCompletedUnits "completed_units"
+
+namespace rstudio {
+namespace session {
+namespace modules {
+namespace rmarkdown {
+namespace notebook {
+
+NotebookDocQueue::NotebookDocQueue(const std::string& docId, 
+      const std::string& jobDesc, CommitMode commitMode,
+      int pixelWidth, int charWidth) :
+      docId_(docId),
+      jobDesc_(jobDesc),
+      commitMode_(commitMode),
+      pixelWidth_(pixelWidth),
+      charWidth_(charWidth)
+{
+}
+
+NotebookDocQueue::NotebookDocQueue()
```

Comment:
That's a good point. At the moment there's no codepath which would result in an empty object, but I think it's better just to eliminate the possibility than to add test methods and guards. I've removed this constructor and factored the code that referenced it so that it can't construct an uninitialized queue.